native/frontend/src/org/jetbrains/kotlin/resolve/konan/diagnostics/NativeObjCNameChecker.kt

```diff
@@ -0,0 +1,145 @@
+        val objCNames = descriptor.getObjCNames().filterNotNull()
+        if (objCNames.isEmpty()) return
+        if (descriptor is CallableMemberDescriptor && descriptor.overriddenDescriptors.isNotEmpty()) {
+            objCNames.forEach {
+                val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(it.annotation) ?: declaration
+                context.trace.report(ErrorsNative.INAPPLICABLE_OBJC_NAME.on(reportLocation))
+            }
+        }
+        objCNames.forEach { checkObjCName(it, declaration, descriptor, context) }
+    }
+
+    // We only allow valid ObjC identifiers (even for Swift names)
+    private val validFirstChars = ('A'..'Z').toSet() + ('a'..'z').toSet() + '_'
+    private val validChars = validFirstChars + ('0'..'9').toSet()
+
+    private fun checkObjCName(
+        objCName: ObjCName,
+        declaration: KtDeclaration,
+        descriptor: DeclarationDescriptor,
+        context: DeclarationCheckerContext
+    ) {
+        if (objCName.name == null && objCName.swiftName == null) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME.on(reportLocation))
+        }
+        val invalidNameFirstChar = objCName.name?.firstOrNull()?.takeUnless(validFirstChars::contains)
+        val invalidSwiftNameFirstChar = objCName.swiftName?.firstOrNull()?.takeUnless(validFirstChars::contains)
+        val invalidFirstChars = setOfNotNull(invalidNameFirstChar, invalidSwiftNameFirstChar)
+        if (invalidFirstChars.isNotEmpty()) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME_FIRST_CHAR.on(reportLocation, invalidFirstChars.joinToString("")))
+        }
+        val invalidNameChars = objCName.name?.toSet()?.subtract(validChars) ?: emptySet()
+        val invalidSwiftNameChars = objCName.swiftName?.toSet()?.subtract(validChars) ?: emptySet()
+        val invalidChars = invalidNameChars + invalidSwiftNameChars
+        if (invalidChars.isNotEmpty()) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INVALID_OBJC_NAME_CHARS.on(reportLocation, invalidChars.joinToString("")))
+        }
+        if (objCName.exact && descriptor !is ClassDescriptor) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.INAPPLICABLE_EXACT_OBJC_NAME.on(reportLocation))
+        }
+        if (objCName.exact && objCName.name == null) {
+            val reportLocation = DescriptorToSourceUtils.getSourceFromAnnotation(objCName.annotation) ?: declaration
+            context.trace.report(ErrorsNative.MISSING_EXACT_OBJC_NAME.on(reportLocation))
```

Comment:
Why?
I mean, `exact = true` affects even Swift name for nested classes. So this restriction prohibits some legitimate case (very corner one though), and the limitation seems a bit unmotivated.

Reply:
Correct `exact = true` also affects the Swift name. However if `name == null` the meaning of `exact = true` doesn't make sense IMO. 
```kotlin 
class Foo {
    @ObjCName(swiftName = "SwiftBar", exact = true)
    class Bar {

    }
}
```
This will result in `Bar` being called `SwiftBar` in Swift, nothing special.
Now would `Bar` be called `Bar` or `FooBar` in ObjC?

Reply:
> Now would `Bar` be called `Bar` or `FooBar` in ObjC?

Good question. So it makes some sense to prohibit that, just to avoid answering this question, at least for now, right?
On the other hand, how should this work when
```kotlin
class Foo {
    @ObjCName(nane = "ObjCBar", exact = true)
    class Bar
}
```
?
Should it be `Foo.Bar` or simply `Bar` in Swift?

Reply:
> Should it be Foo.Bar or simply Bar in Swift?

In Swift this will be `ObjCBar` since Swift inherits the ObjC name unless a Swift specific name is provided.